### PR TITLE
DataViews: iterate on list view

### PR DESCRIPTION
**Pull request by oandregal**
Part of https://github.com/WordPress/gutenberg/issues/55083
Related https://github.com/WordPress/gutenberg/pull/56476#discussion_r1413629329

## What?

This PR iterates on the API and design of the list view:

https://github.com/WordPress/gutenberg/assets/583546/31fd1220-15a2-4e28-9e55-da8a61e1d837

## Why?

- API wise, we need to define how the consumer work with previews.
- At the moment, the list view is using the table layout – it needs its own. Target design:

274858864-cf194048-01a3-4410-8025-4ab1f6354fd2

## How?

- The `render` prop of the field API only takes an `item` as parameter, no longer takes the view, mimicking how `getValue` works.
- Let the consumer configure the list view by setting the primary & media fields.
- We don't plan to have a `list` view that works with and without preview, so it's not necessary to export whether the view supports previews or not.

```js
{
  [ LAYOUT_LIST ]: {
    primaryField: 'title',
    mediaField: 'featured-image',
  }
}
```

- The selection affordance is no longer provided by a field (link), but by the view itself. Expose a `onSelectionChange` prop to `DataViews` component that relays the selection to the consumer, to update the preview.

```js
<DataViews
  onSelectionChange={ ( items ) => { /* consumer instantiates the preview * / } }
/>
``` 

## Testing Instructions

- Enable the "new admin view" experiment and visit "Manage all pages" or "Manage all templates".
- Select the "list" layout and interact with it.

## Follow-ups

- Filters: display them condensed, in a dropdown + reduce the width of the view frame.
- Button to trigger the page details (chrevon right in list items).
- Action items: could be added after the buttor to trigger the page details (see [mockup](https://github.com/WordPress/gutenberg/pull/56746#issuecomment-1840500851)).
- Keyboard navigation: potentially use arrow up/down to navigate through the list (to revisit when the items have buttons within).
- Preview should be a toggle-able field.
- Enable list view for templates.
- Clicking the Frame should go to edit view.

### CODE REVIEW

**Comment by oandregal** [comment]
@jameskoster @SaxonF I've started looking into updating a bit the list view design. Is there any mockup available?

File: packages/edit-site/src/components/page-pages/index.js
Comment (oandregal) [inline packages/edit-site/src/components/page-pages/index.js]: This prop only makes sense if the list view can work with and without previews. Otherwise, we can remove it.

File: packages/dataviews/src/constants.js
Comment (oandregal) [inline packages/dataviews/src/constants.js:32]: This means there is no longer a need to export anything from dataviews (`VIEW_LAYOUTS` or any specific `getViewSupport` function).

File: packages/edit-site/src/components/page-pages/index.js
Comment (oandregal) [reply packages/edit-site/src/components/page-pages/index.js]: Having a preview influences the UI design as well: with preview, the filters are condensed into a single dropdown.

Comment (oandregal) [reply packages/edit-site/src/components/page-pages/index.js]: I ran this by some people, and the common believe is that the list view doesn't make much sense without the preview. For that kind of "row condensed" layout, the table would be better. I'm removing this.

**Comment by oandregal** [comment]
Changes in the last iteration:

- Displayed a media field or placeholder if it doesn't render anything.
- Displayed the visible fields.
- Remove the chevronRight icon: it is meant to open the page details, not to serve as a trigger affordance for the preview. This is to be explored in a follow-up. 

https://github.com/WordPress/gutenberg/assets/583546/8887f8db-c3ce-45a6-8a1c-7d9412591586

File: packages/dataviews/src/view-grid.js
Comment (oandregal) [inline packages/dataviews/src/view-grid.js:46]: We can access the `view` in the fields API directly, no need to pass it to the field through `render`. Same for `view-table.js`. This change is not necessary for this PR to work.

File: packages/dataviews/src/view-list.js
Comment (oandregal) [inline packages/dataviews/src/view-list.js]: @andrewhayward what would be the recommended way to make this work? Either making `li` interactive or substituting it by something else?

**Comment by oandregal** [comment]
I've noted in the description some things I'd like to leave up for follow-ups:

- Filters: display them condensed, in a dropdown.
- Action items.
- Button to trigger the page details. See the chrevon right in list items:

Screenshot 2023-12-05 at 10 46 39

**Comment by youknowriad** [comment]
I think for the "buttons" to go to "details" we should hold on implementing that until we merge the "pages list" with the parent "pages list" page in the site editor. We don't want to introduce "two different details" pages.

**Comment by jameskoster** [comment]
I noticed the list view isn't enabled for templates, any reason why?

I think preview (featured image) should be a toggle-able field.

Would it be a lot of extra work to make the Frame act as a link to edit view? It's a bit strange how you see all these Snackbars when you click it?

Screenshot 2023-12-05 at 11 14 32

There are some general styling adjustments to make, I'll push some changes shortly.

**Comment by jameskoster** [comment]
> There are some general styling adjustments to make, I'll push some changes shortly.

One item I'll need some help with is making the list view column 360px wide (to match the main sidebar). I couldn't see a classname to use for that.

Another one is that the selected item should have an active class / style.

**Comment by jameskoster** [comment]
Another point (sorry for the barrage) – the title should not be a link to editing in this view, just a regular heading.

**Comment by youknowriad** [comment]
I think we should try to note all these in an issue or the overview issue, I'll defer to @oandregal but not everything might be solvable here and this is a great improvement PR.

**Comment by jameskoster** [comment]
I don't mind either way, happy to add notes  👍  The column width issue is the most important I'd say. It's a bit strange how the Frame / list view contains change width as you paginate.

File: packages/dataviews/src/view-list.js
Comment (oandregal) [reply packages/dataviews/src/view-list.js]: I went with https://github.com/WordPress/gutenberg/pull/56746/commits/d0aaf512bef7a5fd49f86650e1d7cb94f79687bf.

(It's not shown in the video but I pressed the `ENTER` key to trigger the preview):

https://github.com/WordPress/gutenberg/assets/583546/28973d9d-1660-4226-be14-42a49d4c4048

File: packages/edit-site/src/components/page-pages/index.js
Comment (oandregal) [inline packages/edit-site/src/components/page-pages/index.js:196]: This switches the current code logic but maintains behavior: for the `grid` view, try finding the largest size. For any other view, do the opposite.

Note that when this code was [introduced](https://github.com/WordPress/gutenberg/pull/55353/files#diff-44881c176d8762d8e2b3cb7d8a82d1fe82c69002b50817de1663052ec167a618R98) the current `table` view was named `list`. This string was missed when the views were renamed, should have been `table` now. So this also fixes a bug in `trunk`.

File: packages/dataviews/src/view-table.js
Comment (oandregal) [inline packages/dataviews/src/view-table.js:347]: We can access the view in the fields API directly, no need to pass it to the field through render. Same for `view-grid.js`. This change is not necessary for this PR to work.

File: packages/dataviews/src/dataviews.js
Comment (oandregal) [inline packages/dataviews/src/dataviews.js:33]: The selection state could have been stored as part of the ViewList component. By managing it at this level, layout switches would maintain selection when the other views implement it. See [related conversation](https://github.com/WordPress/gutenberg/pull/56476#discussion_r1413629329).

File: packages/dataviews/src/view-list.js
Comment (andrewhayward) [inline packages/dataviews/src/view-list.js]: Have you tried/considered using [`Button`](https://wordpress.github.io/gutenberg/?path=/docs/components-button--docs) here? Generally better to use existing components rather than reimplementing behaviour.

Comment (oandregal) [reply packages/dataviews/src/view-list.js]: Yeah. Other than styling, my main concern was that this component is going to have buttons within ([this](https://github.com/WordPress/gutenberg/pull/56746#issuecomment-1840500851), but also actions).

**Comment by andrewhayward** [comment]
Thanks for working on this, @oandregal. I left a comment about using `Button` for the interactive items, rather than reimplementing the functionality. If we don't use `Button`, we should at least make sure that each item has a focus ring, and that [the `onEnter` key handler](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#keyboard_interactions) processes <kbd>Space</kbd> as well as <kbd>Enter</kbd>.

Outside of that, we should set [`aria-pressed`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed) to `true`/`false` on each item, depending on whether the relevant side panel is showing, to indicate that they are toggle buttons.

We should also set [`aria-labelledby`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby) on each item pointing to the `primaryField`, and possibly [`aria-describedby`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) pointing to the rest of the content. At the moment, the buttons are very noisy.

**Comment by andrewhayward** [comment]
Also looks like #56320 could potentially cause issues here too. `primaryField` probably shouldn't be an `H3`.

**Comment by oandregal** [comment]
Thanks for taking a look, @andrewhayward !

> Thanks for working on this, @oandregal. I left a comment about using Button for the interactive items, rather than reimplementing the functionality. If we don't use Button, we should at least make sure that each item has a focus ring, and that [the onEnter key handler](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#keyboard_interactions) processes Space as well as Enter.
> 
> Outside of that, we should set [aria-pressed](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed) to true/false on each item, depending on whether the relevant side panel is showing, to indicate that they are toggle buttons.

I've done all of this. Note what I've mentioned about using [buttons](https://github.com/WordPress/gutenberg/pull/56746#discussion_r1415880149).

> We should also set [aria-labelledby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby) on each item pointing to the primaryField, and possibly [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) pointing to the rest of the content. At the moment, the buttons are very noisy.
> Also looks like https://github.com/WordPress/gutenberg/issues/56320 could potentially cause issues here too. primaryField probably shouldn't be an H3.

These two are related. The gist of it is that the field that acts as a primary would need to provide some sort of handle, or let the view do that. I could make it work for this _page_, though I'd rather investigate this separately at make it work for all _dataviews_. Would that be ok?

File: packages/dataviews/src/view-list.js
Comment (andrewhayward) [reply packages/dataviews/src/view-list.js]: Hmm... OK. In which case, [we shouldn't be doing this at all](https://accessibleweb.com/question-answer/why-are-nested-interactive-controls-an-accessibility-issue/#:~:text=When%20an%20interactive%20control%20has,information%2C%20or%20interacting%20with%20it.)! Will need to have a think about how best to handle this scenario.

File: packages/dataviews/src/dataviews.js
Comment (youknowriad) [reply packages/dataviews/src/dataviews.js:33]: Is it a bad thing to retain selection when switching views?

File: packages/dataviews/src/view-list.js
Comment (jameskoster) [reply packages/dataviews/src/view-list.js]: Hopefully it's okay to 'visually nest' those buttons, with either absolute positioning, something like the List view implementation, or some other technique?

**Review by jameskoster — APPROVED** [review]
Generally speaking I think the design is in a decent spot. There are some follow-ups considered too large to address this PR as noted in https://github.com/WordPress/gutenberg/issues/55083.

https://github.com/WordPress/gutenberg/pull/56786 will also help to tidy up some of the layout quirks.

File: packages/dataviews/src/view-list.js
Comment (andrewhayward) [reply packages/dataviews/src/view-list.js]: > Hopefully it's okay to 'visually nest' those buttons, with either absolute positioning, something like the List view implementation, or some other technique?

Yeah, it's a pretty well established paradigm to visually overlay interactive elements. We just can't nest them in the DOM. For example, this card pattern is routinely used on news sites and in other related content...

<p align="center">
  An article 'card' showing an image, a large title, some text, and a category. The title is underlined with a dark focus ring.
  An article 'card' showing an image, a large title, some text, and a category. The category is underlined with a dark focus ring.
</p>

There are two interactive elements in this card – links for the article title, and the the article category.

<p align="center">
  An article 'card' showing an image, a large title, some text, and a category. The title is underlined, and a mouse pointer is visible in the centre of the card.
  An article 'card' showing an image, a large title, some text, and a category. The category is underlined, with a mouse pointer visible on top of it.
</p>

However, through some clever use of [generated content](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_generated_content), and a bit of `z-index` magic, the entire card becomes clickable, and behaves as if it's all one big link, unless you intentionally hover over the category link, which then becomes the target action.

We don't have to implement it exactly like that, but the general idea is there. One thing to be aware of when doing this is target size – it's bad enough when it's hard to tap on something, but when it's surrounded by a different action context it's even worse.

File: packages/edit-site/src/components/page-pages/index.js
Comment (ntsekouras) [inline packages/edit-site/src/components/page-pages/index.js:53]: Maybe the `primaryField` should be shared by all layouts.. That's a discussion that's not blocking this PR of course..

**Review by ntsekouras — APPROVED** [review]
Thanks @oandregal! This is definitely a good improvement and I think we can land and iterate.

File: packages/dataviews/src/dataviews.js
Comment (oandregal) [reply packages/dataviews/src/dataviews.js:33]: Oh, it's good, I think. What I meant is that I could have stored state in the `ViewList` in this PR, though I had it in `DataViews` to support that (retain selection upon layout switching).

File: packages/dataviews/src/view-list.js
Comment (oandregal) [reply packages/dataviews/src/view-list.js]: Nice, thanks for sharing the details. It'll be useful when we implement this feature.

**Comment by oandregal** [comment]
Oh, I think this got merged before all tests were green: the react native test was still running.

I set up the "auto-merge" and, upon accepting, the PR got merged directly – without waiting for the test to end. I don't think auto-merge skips the react native tests? I also don't think I've unintentionally clicked the "use admin permissions" to merge directly? I don't see how this happened. 🤔

Anyway, looking at the [test failure](https://github.com/WordPress/gutenberg/actions/runs/7103878062/job/19402690513), it's unrelated to this. I'll keep an eye on `trunk` runs.

Edit: `trunk` is green.

**Comment by ciampo** [comment]
> > We should also set [aria-labelledby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby) on each item pointing to the primaryField, and possibly [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) pointing to the rest of the content. At the moment, the buttons are very noisy.
> > Also looks like #56320 could potentially cause issues here too. primaryField probably shouldn't be an H3.
> 
> These two are related. The gist of it is that the field that acts as a primary would need to provide some sort of handle, or let the view do that. I could make it work for this _page_, though I'd rather investigate this separately at make it work for all _dataviews_. Would that be ok?

@andrewhayward ,  Is it worth moving this conversation to a new issue to track it separately?

And maybe [this one](https://github.com/WordPress/gutenberg/pull/56746#discussion_r1415880149) too ?

**Comment by oandregal** [comment]
> > We should also set [aria-labelledby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby) on each item pointing to the primaryField, and possibly [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby) pointing to the rest of the content. At the moment, the buttons are very noisy.
> Also looks like https://github.com/WordPress/gutenberg/issues/56320 could potentially cause issues here too. primaryField probably shouldn't be an H3.
> 
> These two are related. The gist of it is that the field that acts as a primary would need to provide some sort of handle, or let the view do that. I could make it work for this page, though I'd rather investigate this separately at make it work for all dataviews. Would that be ok?

I'm exploring a potential path for this https://github.com/WordPress/gutenberg/pull/56942

**Comment by oandregal** [comment]
I went through all this PR and updated the follow-up tasks created by Jay to the tracking issue https://github.com/WordPress/gutenberg/issues/55083 Please, take a look and share if I've missed anything.

**Pull request merged**
